fecgov/fec-cms
# [WIP] Update d3 version, includes, and usage—initial commit

## Summary
We're still using v3 of d3.js and would like to upgrade to v5. The step from v3 to v4 was significant, though, so we'd like to thoroughly test functionality before we do.

- Resolves # [2668](https://github.com/fecgov/fec-cms/issues/2668)

## Impacted areas of the application
Files affected:
- fec/fec/static/js/modules/line-chart.js
- fec/fec/static/js/modules/line-chart-committees.js
- fec/fec/static/js/modules/maps.js
- (package.json, etc)

Changes:
- Installed the newest version of d3 (5.8.2)
- Updated `require` commands to include only the modules used for those specific files, rather than including the entire d3 package*
- Addressed general naming updates (i.e., `d3.time.format` => `d3.timeFormat`)
- Addressed generally more simplified d3 functionality (i.e., `xAxis = d3.axis().scale(x).orient('bottom') => `xAxis = d3.axisBottom(x)`)
- Addressed new complications with packaged d3 and events (seems Webpack, etc. tend to confuse d3.event and native browser events)

* current files size decrease estimates:
`data-browse-data-*`: 2.14 MB => 925 kB (**-1.215 MB**)
`data-landing-*`: 4.91 MB => 3.68 MB (**-1.23 MB**)
`dataviz-common-*`: 7.48 MB => 5.85 MB (**-1.63 MB**)

## Screenshots
![image](https://user-images.githubusercontent.com/26720877/52424612-5b979f80-2ac8-11e9-8dfb-a2d1b659a785.png)
![image](https://user-images.githubusercontent.com/26720877/52424722-8aae1100-2ac8-11e9-9637-693ecb562a17.png)
![image](https://user-images.githubusercontent.com/26720877/52424797-b7622880-2ac8-11e9-8384-4c0ce6e8ba8f.png)
![image](https://user-images.githubusercontent.com/26720877/52424870-dc569b80-2ac8-11e9-9722-ae1cd9c5d37a.png)


## How to test
First, upgrade the installed version of d3.js. I used `npm i --save d3@latest`
Double check that the `FEC_FEATURE_LINECHARTS` feature flag is `true`
- Check the map on [localhost](http://localhost:8000/data) for correct functionality (click, zoom, etc) when compared to [Production](https://www.fec.gov/data/)
- Check the line chart on [localhost](http://localhost:8000/data) for correct functionality (mouseover the lines, click the arrows) when compared to [Production](https://www.fec.gov/data/)
- Compare the Raising charts for [localhost](http://localhost:8000/data/browse-data/?tab=raising) and [Production](https://www.fec.gov/data/browse-data/?tab=raising)
- Compare the Spending charts for [localhost](http://localhost:8000/data/browse-data/?tab=spending) and [Production](https://www.fec.gov/data/browse-data/?tab=spending)

____
